cpp/src/community/leiden_kernels.cu

```diff
@@ -0,0 +1,289 @@
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename edge_t, typename weight_t>
+weight_t update_clustering_by_delta_modularity_constrained(
+  weight_t total_edge_weight,
+  weight_t resolution,
+  GraphCSRView<vertex_t, edge_t, weight_t> const &graph,
+  rmm::device_vector<vertex_t> const &src_indices,
+  rmm::device_vector<weight_t> const &vertex_weights,
+  rmm::device_vector<weight_t> &cluster_weights,
+  rmm::device_vector<vertex_t> &cluster,
+  rmm::device_vector<vertex_t> &constraint,
+  cudaStream_t stream)
+{
+  rmm::device_vector<vertex_t> next_cluster(cluster);
+  rmm::device_vector<weight_t> delta_Q(graph.number_of_edges);
+  rmm::device_vector<vertex_t> cluster_hash(graph.number_of_edges);
+  rmm::device_vector<weight_t> old_cluster_sum(graph.number_of_vertices);
+
+  weight_t *d_delta_Q           = delta_Q.data().get();
+  vertex_t *d_constraint        = constraint.data().get();
+  vertex_t const *d_src_indices = src_indices.data().get();
+  vertex_t const *d_dst_indices = graph.indices;
+
+  weight_t new_Q = modularity(total_edge_weight, resolution, graph, cluster.data().get(), stream);
+
+  weight_t cur_Q = new_Q - 1;
+
+  // To avoid the potential of having two vertices swap clusters
+  // we will only allow vertices to move up (true) or down (false)
+  // during each iteration of the loop
+  bool up_down = true;
+
+  while (new_Q > (cur_Q + 0.0001)) {
+    cur_Q = new_Q;
+
+    compute_delta_modularity(total_edge_weight,
+                             resolution,
+                             graph,
+                             src_indices,
+                             vertex_weights,
+                             cluster_weights,
+                             cluster,
+                             cluster_hash,
+                             delta_Q,
+                             old_cluster_sum,
+                             stream);
+
+    // Filter out positive delta_Q values for nodes not in the same constraint group
+    thrust::for_each(
+      rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_edges),
+      [d_src_indices, d_dst_indices, d_constraint, d_delta_Q] __device__(vertex_t i) {
+        vertex_t start_cluster = d_constraint[d_src_indices[i]];
+        vertex_t end_cluster   = d_constraint[d_dst_indices[i]];
+        if (start_cluster != end_cluster) d_delta_Q[i] = weight_t{0.0};
+      });
+
+    assign_nodes(graph,
+                 delta_Q,
+                 cluster_hash,
+                 src_indices,
+                 next_cluster,
+                 vertex_weights,
+                 cluster_weights,
+                 up_down,
+                 stream);
+
+    up_down = !up_down;
+
+    new_Q = modularity(total_edge_weight, resolution, graph, next_cluster.data().get(), stream);
+
+    if (new_Q > cur_Q) { thrust::copy(next_cluster.begin(), next_cluster.end(), cluster.begin()); }
```

Comment:
Consider using `thrust::copy(rmm::exec_policy(stream)->on(stream),...)` that carries stream information.

Reply:
raft also has copy, update_host, update_device, ... taking stream.

https://github.com/rapidsai/raft/blob/branch-0.15/cpp/include/raft/cudart_utils.h#L199

Reply:
Added the exec_policy.  raft utilities seem to want raw pointers, which we don't have here.